Add Term Merge module

mstenta opened the pull request:
This adds the Drupal Term Merge module to farmOS, along with a simple `farm_ui_term` module that depends on it and adds the necessary permission.

For more information, see this forum topic: https://farmos.discourse.group/t/merge-module-testing/1635

paul121 commented:
I don't have a strong opinion on this.  It does sound useful, but only once your farmOS instance gets "messy" :-) As such I do think it would be nice if this was an optional module and not always enabled with `farm_ui`. Otherwise it is just more "things" clutter/etc for new users to learn. :shrug:

mstenta commented:
> it would be nice if this was an optional module and not always enabled with `farm_ui`.

I thought about this too.

We need to wrap it in a "ui" module either way, so that we can add the necessary permission. Since the permission is the only bit that's needed (on top of enabling the module), I originally thought about just sticking it into an existing UI module, but decided it made sense to keep as it's own "term UI" module. This gives site admins the option to remove it if they want. So the flexibility is there, it's just a question of whether or not it should be installed by default.

If we don't, then we need to add it to `setup/modules` so that non-admins can find/enable it.

> Otherwise it is just more "things" clutter/etc for new users to learn.

My feeling is that this is a small enough "thing" that it isn't a burden, and it's purpose is intuitively clear. All it adds is a "Merge" tab to the term list page (where you would intuitively look for it).

![Screenshot from 2025-05-28 06-58-16](https://github.com/user-attachments/assets/fa95e796-87b2-48e5-9685-8fbb5ac0b837)

![Screenshot from 2025-05-28 06-58-30](https://github.com/user-attachments/assets/5a92777c-a7b5-4319-a19d-aa6f063dc5f6)

This is the sort of thing that you don't know you need until you need it, and is so small that you might not think to look for it as an optional module. IMO someone who finds themselves with a cluttered taxonomy would look for some kind of option in the Setup > Taxonomy section.

I think the intuitive feature benefit outweighs the "clutter" of adding one tab. Honestly, this feels to me like a feature that should be part of Drupal core taxonomy UI.

> It does sound useful, but only once your farmOS instance gets "messy" :-)

I think this is a very common problem. Whether due to user error or bugs in code, I've seen exact duplicate terms in almost every site I've looked at - especially in the Units vocabulary (eg: multiple terms named "lbs"). It could be that there were bugs in the past that created them, or bugs right now that we're unaware of (or *future* bugs in core/contrib modules!). Those would be separate issues in themselves, of course, but IMO there's a need to provide some more control to help users manage their vocabularies, even if it's not about exact duplicates. The other thing I see all the time are terms that *should* be merged (like "lb", "lbs", "pound", "pounds").

Without this module there is no way to deal with these, other than manually cleaning them up (and ALL the entities that reference them). I've personally experienced what it feels like to realize that - it's very frustrating and disheartening! And because the burden of cleaning them up manually is so high, most users will simply live with the mess (myself included)! Over time that mess wears on you. :sweat_smile:

So in my mind this is a very small addition (a single "Merge" tab, low/no maintenance burden) with a very big UX benefit.

mstenta commented:
Rebased onto `4.x`, updated for Drupal 11, and changed the `drupal/term_merge` version constraint from `^2.0@beta` to `^2.0`. I also tested the update hook and term merge module manually again. All is working as expected.
